docs: BFF ADR (#1238)
openedx/frontend-app-learner-portal-enterprise: docs/decisions/0013-backend-for-frontend-api.md

# 0013. Introduction of the backend-for-frontend orchaestration API

## Status

Accepted (December 2024)

## Context

As part of supporting the feature of auto enrollment of default enrollment intentions ([see ADR](https://github.com/openedx/edx-enterprise/blob/master/docs/decisions/0015-default-enrollments.rst))
and auto-apply of subscription plan licenses, it was determined the sequencing of API calls to resolve in the existing 
frontend root loader would result in a new set of waterfall API calls on top of existing reconciliation of the current status of

> the user's subsidies when arriving on the enterprise learner portal. Furthermore, in support of including this feature,

Reviewer: Bit of a correction here: the primary concern is the interaction of API requests between the `rootLoader` and `dashboardLoader`, as they operate asynchronously in parallel. The challenge presents itself in that the side effect of realizing default enrollment intentions would need to happen in the `rootLoader` after processing the auto-applied subscription license, which would effectively change the response of the enrollments retrieved by the `dashboardLoader`.

Another way of putting it might be that, currently the `rootLoader` and `dashboardLoader` operate independently as neither mutates the data for the other. To support auto-enrollment, though, the `rootLoader` would mutate the data retrieved by `dashboardLoader`, which would lead to additional complexities around data orchestration regarding optimistic cache updates and possible race conditions.

it required additional business logic to be parsed about the current state of the requester's association to the enterprise,

> the current state of their subsidies (and auto-apply of subscription licenses), and the eventual auto enrollment of a 
> default enrollment intention.

Reviewer: nit: 

```suggestion
the current state of their subsidies (and auto-apply of subscription licenses), and the eventual realization of a 
default enrollment intention.
```

> 

Reviewer: It might also be worth expanding on the long-term performance rationale for wanting to abstract the API orchestration from the frontend in favor of a primary single API endpoint per page route, starting with the Dashboard.

## Decision

The recognition of the additional maintenance of business logic with the default enrollment flow along with the overhead
of resolving the series of API calls required to resolve the requirements of the feature resulted in the creation of a 

> backend-for-frontend (BFF) layer within the enterprise-access service ([see ADR](https://github.com/openedx/enterprise-access/blob/main/docs/decisions/)).

Reviewer: nit: 

```suggestion
Backend-for-Frontend (BFF) / API Gateway layer within the enterprise-access service ([see ADR](https://github.com/openedx/enterprise-access/blob/main/docs/decisions/)).
```

> backend-for-frontend (BFF) layer within the enterprise-access service ([see ADR](https://github.com/openedx/enterprise-access/blob/main/docs/decisions/)).

Reviewer: nit: feel free to leave off a link to the related ADR for now; I think this ADR will be ready sooner than that one; when it's up, can open another PR against this ADR to add the relevant link.

> A BFF layer is a accepted architectural consideration when designing for microservices when the consideration of consolidating

Reviewer: nit:

```suggestion
A BFF layer is an accepted architectural pattern when designing for microservices when the consideration of consolidating
```

business logic dependent on multiple services and an intentional reduction of latency between N number of services required to 
determine the resultant business logic outcome versus a single API call with the logic already resolved to be used.

For the enterprise learner portal implementation, currently, the dashboard route, `/:enterpriseSlug`, is the only route currently

> implementing the BFF layer on _specific_ API calls for the sole intention of supporting, auto apply licenses and auto enrollment via default intentions.

Reviewer: This paragraph could be expanded to inform that the goal/intent is incrementally migrate route-by-route, starting with the Dashboard route as you have mentioned already.

> For the enterprise learner portal implementation, currently, the dashboard route, `/:enterpriseSlug`, is the only route currently
> implementing the BFF layer on _specific_ API calls for the sole intention of supporting, auto apply licenses and auto enrollment via default intentions.

Reviewer: nit: 

```suggestion
For the enterprise learner portal implementation, currently, the dashboard route, `/:enterpriseSlug`, is the only route currently
implementing the BFF layer on _specific_ API calls for the sole intention of supporting, auto-apply licenses, and realizing default enrollment intentions.
```

### Making the BFF API call via `useBFF`

> At a high level, the implications to the frontend is the inclusion of the following hooks, services, and keys and constraints: 

Reviewer: nit: 

```suggestion
At a high level, the implications to the frontend is the inclusion of the following hooks, services, query keys, and constraints: 
```

- `fetchEnterpriseLearnerDashboard` ([source](https://github.com/openedx/frontend-app-learner-portal-enterprise/blob/337a7b44d94d8be5d9233e80e2fa0e2de72d165c/src/components/app/data/services/bffs.ts))

>   - This is the function that makes the service call to the BFF. It requires either a `entepriseSlug` or `enterpriseUuid` to resolve successfully. 

Reviewer: nit: maybe the following?

```suggestion
  - This is the function that makes the service call to the Dashboard BFF API. It requires either a `entepriseSlug` or `enterpriseUuid` to resolve successfully. 
```

>   - This is the function that makes the service call to the BFF. It requires either a `entepriseSlug` or `enterpriseUuid` to resolve successfully. 

Reviewer: nit: might consider removing the latter sentence about the "It require either a slug or uuid"; I think we should consider that long-term only the slug would ideally be used in practice, even if the function technically supports both, since the UUID will eventually be coming from the BFF response, derived from the slug.

> - Creation of the query key ([source](https://github.com/openedx/frontend-app-learner-portal-enterprise/blob/337a7b44d94d8be5d9233e80e2fa0e2de72d165c/src/components/app/data/queries/queryKeyFactory.js#L267))

Reviewer: nit: it seems the pattern in these list items is to include the function name first, followed by the description. for example, the line item above starts with `fetchEnterpriseLearnerDashboard` and the line item below starts with `useBFF`. Should this line item start with `queryEnterpriseLearnerDashboardBFF`?

  and helper function `queryEnterpriseLearnerDashboardBFF` ([source](https://github.com/openedx/frontend-app-learner-portal-enterprise/blob/337a7b44d94d8be5d9233e80e2fa0e2de72d165c/src/components/app/data/queries/queries.ts#L270))
- `useBFF` ([source](https://github.com/openedx/frontend-app-learner-portal-enterprise/blob/337a7b44d94d8be5d9233e80e2fa0e2de72d165c/src/components/app/data/hooks/useBFF.js#L16))

>   - Since the introduction of React Query and custom hooks, the migration process involved abstracting a generalized `useBFF` hook
>     that determines the following conditionals:

Reviewer: [suggestion] 

```suggestion
  - Since the introduction of React Query and custom hooks, the migration process involved abstracting a generalized `useBFF` hook to aid in the incremental migration to the BFF APIs away from the current service requests:
```

    - Within `resolveBFFQuery` ([source](https://github.com/openedx/frontend-app-learner-portal-enterprise/blob/337a7b44d94d8be5d9233e80e2fa0e2de72d165c/src/components/app/data/queries/utils.js#L11))
      - Whether a customer is eligible to use the BFF API call by either being explicitly enabled via the configuration API as a defined `enterpriseUuid`,

>         or an individual customer has been selected to resolve the dashboard page with the BFF loader with a gradual rollout process via a Waffle flag.

Reviewer: ```suggestion
        or an individual enterprise learner has been selected to resolve the dashboard page with the BFF loader with a gradual rollout process via a Waffle flag exposes via `enterpriseCustomer.enterpriseFeatures`.
```

>         or an individual customer has been selected to resolve the dashboard page with the BFF loader with a gradual rollout process via a Waffle flag.

Reviewer: nit: is it worth expanding on this slightly to inform the feature flagging is temporary only to support the mentioned gradual rollout of the BFF to all learners/customers?

>       - Whether the current route the user is on is eligible to make the BFF call, if so, return the query required to call the BFF.

Reviewer: [suggestion] might be worth expanding on what "eligible" means in this case, i.e. the page route has a corresponding BFF API endpoint that exists.

>     - For a matched route where a BFF call can be made, but the user is ineligible to make the call, we fall back to the default query provided by a

Reviewer: ```suggestion
    - For a matched route where a BFF call can be made, but the user is ineligible to make the call, we fallback to the default existing query provided by a
```

>       an argument

Reviewer: nit:

```suggestion
      the option `fallbackQueryConfig`
```

        - This is the simplest implementation of migrating existing React query custom hooks to utilize the BFF. It passes an object representing 
          fields typically passed within `useQuery` for the BFF query call within `bffQueryConfig`. This includes the transform to match the
          data structure of the existing `useSubscriptions` non-BFF call to `useBFF` within an object when making a call to the BFF query. It also passes
          the original arguments for making a query call within `useSubscriptions` passed within `useQuery` as a `fallbackQueryConfig` object within `useBFF`.
        - This structure allows the developer to continue using the original hooks as implemented.
      - `useEnterpriseCourseEnrollments` ([source](https://github.com/openedx/frontend-app-learner-portal-enterprise/blob/337a7b44d94d8be5d9233e80e2fa0e2de72d165c/src/components/app/data/hooks/useEnterpriseCourseEnrollments.js#L36))
        - The modifications required to make this current hook utilize the original `useEnterpriseCourseEnrollments` hook and

>           the `useBFF` layer required specific options to be passed in depending on the modified hooks downstream related to `enterpriseCourseEnrollments`.

Reviewer: These hooks appear to be siblings of `resolveBFFQuery`, as a child under `useBFF`, despite these being their own hooks that _consume_ `useBFF`. Should these should be siblings to `useBFF` instead, as they're not part of `useBFF` as `resolveBFFQuery` is?

Perhaps there's a new group as sibling of `useBFF` with the label of something like "Custom query hooks" and a sentence or two description pertaining to "supporting the incremental migration to the BFF layer without impacting existing query hooks to retrieve data throughout the application".

Then, the specifics about `useSubscriptions` and `useEnterpriseCourseEnrollments` could be child items under that parent "Custom query hooks"?

### Incremental migration of existing services to the BFF

In the long run, as additional services currently being called in the frontend will eventually be migrated to the BFF

> layer. There is no definitive timeline of the migration process, but the intention is to provide a seamless experience

Reviewer: grammar nit: should the word "as" above be removed to make the first sentence read better? As is, it seems the sentence kind of cuts off after the "as additional services" clause.

> for feature development by developers and an enjoyable customer experience with continual improvement. 

Reviewer: ```suggestion
for feature development by developers with minimal impact to UI development using existing conventions and an enjoyable customer experience with continual performance improvement. 
```

## Alternatives Considered

The alternative considered was to include the business logic and additional API resolution to return to the user an auto enrolled course

> and an auto-applied license based on just the requester information. This would have resulted in additional latency before loading

Reviewer: nit: what is meant by "requester information"?

> the learner dashboard page adding to an already challenging first contentful paint time based on the additional waterfall request.

Reviewer: ```suggestion
the learner dashboard page adding to an already challenging largest contentful paint time based on the additional waterfall request.
```
